File: Regionales/2013 Northwestern Europe/nwerc-2013-problems-solutions-testcases/NWERC-2013-Problems-Solutions-Testcases/problems/Battle_for_Silver/solutions/tobi.java
// @EXPECTED_RESULTS@: CORRECT
import java.io.*;
import java.util.*;

public class tobi {
	public static final int MAX_V = 6005;
	public static ArrayList<ArrayList<Integer> > adjList;
	public static int silver[] = new int[MAX_V];
	public static boolean dis[] = new boolean[MAX_V];

	public static int chain(ArrayList<Integer> curVs, ArrayList<Integer> sel, int pos, int sum) {
		if (pos >= curVs.size()) return sum;
		int ret = Math.max(sum, chain(curVs, sel, pos+1, sum));
		boolean possible = true;
		for (Integer seli : sel) {
			if (!adjList.get(curVs.get(pos)).contains(seli)) possible = false;
		}
		sel.add(curVs.get(pos));
		if (possible) ret = Math.max(ret, chain(curVs, sel, pos + 1, sum + silver[curVs.get(pos)]));
		sel.remove(sel.size() - 1);
		return ret;
	}

	public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);
		while (sc.hasNext()) {
			int numV, numE, from, to, best = 0;
			numV = sc.nextInt();
			numE = sc.nextInt();
			adjList = new ArrayList<>();
			for (int i = 0; i < numV; i++) {
				adjList.add(new ArrayList<Integer>());
				dis[i] = false;
				silver[i] = sc.nextInt();
				best = Math.max(best, silver[i]);
			}
			for (int i = 0; i < numE; i++) {
				from = sc.nextInt();
				to = sc.nextInt();
				from--; to--;
				adjList.get(from).add(to);
				adjList.get(to).add(from);
			}
			ArrayList<Pair> degrees = new ArrayList<>();
			for (int i = 0; i < numV; i++) {
				degrees.add(new Pair(adjList.get(i).size(), i));
			}
			Collections.sort(degrees, new PairComp());
			for (int i = 0; i < numV; i++) {
				int curV = degrees.get(i).second;
				ArrayList<Integer> curVs = new ArrayList<>();
				for (Integer next : adjList.get(curV)) {
					if (!dis[next]) curVs.add(next);
				}
				ArrayList<Integer> sel = new ArrayList<>();
				best = Math.max(best, chain(curVs, sel, 0, silver[curV]));
				dis[curV] = true;
			}
			System.out.println(best);
		}
	}
}
class PairComp implements Comparator<Pair> {
	public int compare(Pair p1, Pair p2) {
		if (p1 == null && p2 == null) return 0;
		if (p1 == null) return -1;
		if (p2 == null) return 1;
		if (p1.first == p2.first) return p1.second - p2.second;
		return p1.first - p2.first;
	}
}
class Pair {
	int first, second;
	Pair(int f, int s) {
		first = f;
		second = s;
	}
}

File: Regionales/2013 Northwestern Europe/nwerc-2013-problems-solutions-testcases/NWERC-2013-Problems-Solutions-Testcases/problems/Battle_for_Silver/solutions/tobi.cpp
// @EXPECTED_RESULTS@: CORRECT
#include <iostream>
#include <vector>
#include <algorithm>
#define FOR(i,a,b) for (int i = (a); i < (b); i++)
#define contains(a,b) (find((a).begin(), (a).end(), (b)) != (a).end())
#define MAX_V 6005

using namespace std;
vector<int> adjList[MAX_V];
int silver[MAX_V];
bool dis[MAX_V];

int chain(vector<int> curVs, vector<int> sel, int pos, int sum) {
	if (pos >= curVs.size()) return sum;
	int ret = max(sum, chain(curVs, sel, pos+1, sum));
	bool possible = true;
	FOR(i,0,sel.size()) if (!contains(adjList[curVs[pos]], sel[i])) possible = false;
	sel.push_back(curVs[pos]);
	if (possible) ret = max(ret, chain(curVs, sel, pos + 1, sum + silver[curVs[pos]]));
	return ret;
}

int main(int argc, char **argv) {
	while (cin.good() && !cin.eof()) {
		int numV, numE, from, to, best = 0;
		cin >> numV >> numE;
		if (cin.eof()) break; // hate this testcase format
		FOR(i,0,numV) {
			adjList[i].clear();
			dis[i] = false;
			cin >> silver[i];
			best = max(best, silver[i]);
		}
		FOR(i,0,numE) {
			cin >> from >> to;
			from--; to--;
			adjList[from].push_back(to);
			adjList[to].push_back(from);
		}
		vector<pair<int,int> > degrees;
		FOR(i,0,numV) degrees.push_back(make_pair(adjList[i].size(),i));
		sort(degrees.begin(), degrees.end());
		FOR(i,0,numV) {
			int curV = degrees[i].second;
			vector<int> curVs;
			FOR(j,0,adjList[curV].size()) {
				int next = adjList[curV][j];
				if (!dis[next]) curVs.push_back(next);
			}
			vector<int> sel;
			best = max(best, chain(curVs, sel, 0, silver[curV]));
			dis[curV] = true;
		}
		cout << best << endl;
	}
	return 0;
}
